File: src/ui/formatting.py
"""Convert usage values into the strings shown in the interface.

The wording follows Claude's own usage view: a window is named "Current session"
or "This week" and is described by when it resets, as a wall clock time rather
than a countdown. Both the icon tooltip and the panel describe the same numbers,
so that wording lives here once instead of being written twice.
"""

from __future__ import annotations

from datetime import datetime, timedelta, timezone

from ..usage.snapshot import UsageWindow
from ..validation import require_number_in_range, require_type


def format_percent(percent: float) -> str:
	"""Return a usage percentage as a whole number followed by a percent sign.

	Args:
		percent: Share of a window already used, 0 to 100. float.

	Returns:
		str: Text such as "24%".
	"""
	require_number_in_range(percent, 0.0, 100.0, "percent")
	return f"{int(round(percent))}%"


def local_minute(moment: datetime) -> datetime:
	"""Return a moment in local time, rounded to the nearest minute.

	Reset times arrive one second short of the hour, so rounding is what turns
	"Saturday 11:59:59 PM" into the "Sunday 12:00 AM" a reader expects.

	Args:
		moment: The moment to convert. datetime, aware or naive UTC.

	Returns:
		datetime: The same moment in the local time zone, with seconds dropped.
	"""
	require_type(moment, datetime, "moment")
	aware = moment if moment.tzinfo is not None else moment.replace(tzinfo=timezone.utc)
	local = aware.astimezone()
	rounded = local.replace(second=0, microsecond=0)
	return rounded + timedelta(minutes=1) if local.second >= 30 else rounded


def format_clock(moment: datetime) -> str:
	"""Return the local wall clock time of a moment, on a twelve hour clock.

	Args:
		moment: The moment to render. datetime, aware or naive UTC.

	Returns:
		str: Text such as "2:00 PM".
	"""
	local = local_minute(moment)
	return f"{local.hour % 12 or 12}:{local.minute:02d} {'AM' if local.hour < 12 else 'PM'}"


def format_reset_at(resets_at: datetime | None, now: datetime | None = None) -> str:
	"""Return a sentence saying when a window resets.

	The day is named only when the reset is not today, which is how Claude's own
	usage view reads.

	Args:
		resets_at: Moment the window resets, or None when none was reported.
			datetime or None.
		now: Reference moment, or None for the current time. datetime or None.

	Returns:
		str: Text such as "Resets at 2:00 PM" or "Resets Sunday 12:00 AM", and an
		empty string when there is no reset time.
	"""
	if resets_at is None:
		return ""
	require_type(resets_at, datetime, "resets_at")
	if now is not None:
		require_type(now, datetime, "now")

	reference = (now or datetime.now(timezone.utc)).astimezone()
	local = local_minute(resets_at)
	if local.date() == reference.date():
		return f"Resets at {format_clock(resets_at)}"
	return f"Resets {local.strftime('%A')} {format_clock(resets_at)}"


def format_reset(window: UsageWindow, now: datetime | None = None) -> str:
	"""Return a sentence saying when a usage window resets.

	Args:
		window: The window to describe. UsageWindow.
		now: Reference moment, or None for the current time. datetime or None.

	Returns:
		str: Text such as "Resets at 2:00 PM", and an empty string when the
		endpoint reported no reset time.
	"""
	require_type(window, UsageWindow, "window")
	return format_reset_at(window.resets_at, now)


def format_subtitle(window: UsageWindow, now: datetime | None = None) -> str:
	"""Return the second line shown under a window name.

	Args:
		window: The window to describe. UsageWindow.
		now: Reference moment, or None for the current time. datetime or None.

	Returns:
		str: The reset sentence, preceded on its own line for a per model window
		by what that limit covers.
	"""
	require_type(window, UsageWindow, "window")
	reset = format_reset(window, now)
	if not window.scope_name:
		return reset
	scope = f"Separate weekly limit for {window.scope_name}"
	return f"{scope}\n{reset}" if reset else scope


def format_next_update(seconds: float) -> str:
	"""Return the line saying how long until the next reading.

	Args:
		seconds: Time remaining until the next reading. float, negative values
			read as zero.

	Returns:
		str: Text such as "Next update in 42s".
	"""
	require_type(seconds, (int, float), "seconds")
	return f"Next update in {max(0, int(seconds))}s"

File: src/mac/FORL/Core/Formatting.swift
//
//  Formatting.swift
//  Turn usage values into the strings every surface shows.
//
//  The wording follows Claude's own usage view: a window is named "Current
//  session" or "This week" and is described by when it resets, as a wall clock
//  time rather than a countdown. Every surface describes the same numbers, so
//  the wording lives here once.
//

import Foundation

/// The strings a reading is written as.
enum Formatting {
	/// Return a usage percentage as a whole number followed by a percent sign.
	///
	/// - Parameter percent: Share of a window already used, 0 to 100.
	/// - Returns: Text such as "24%".
	static func percent(_ percent: Double) -> String {
		"\(Int(clampPercent(percent).rounded()))%"
	}

	/// Return the hover text for one small surface.
	///
	/// The reading comes first, because how much is left is the question a hover
	/// asks, and the name follows it, because several items may be up at once
	/// and a number alone does not say which limit it belongs to.
	///
	/// - Parameters:
	///   - value: Share of the window already used, 0 to 100.
	///   - label: Name of the window, such as "Current session".
	///   - resetsAt: Reset time for the reading, or nil when unavailable.
	/// - Returns: The percentage and window name, followed by the reset time when known.
	static func tooltip(value: Double, label: String, resetsAt: Date? = nil) -> String {
		let resetText = reset(resetsAt)
		let title = "\(percent(value)) - \(label)"
		return resetText.isEmpty ? title : "\(title) - \(resetText)"
	}

	/// Return the local wall clock time of a moment, on a twelve hour clock.
	///
	/// - Parameter moment: The moment to render.
	/// - Returns: Text such as "2:00 PM".
	static func clock(_ moment: Date) -> String {
		let formatter = DateFormatter()
		formatter.locale = Locale(identifier: "en_US_POSIX")
		formatter.dateFormat = "h:mm a"
		return formatter.string(from: rounded(moment))
	}

	/// Return a sentence saying when a window resets.
	///
	/// The day is named only when the reset is not today, which is how Claude's
	/// own usage view reads.
	///
	/// - Parameters:
	///   - resetsAt: Moment the window resets, or nil when none was reported.
	///   - now: Reference moment, for deciding whether the reset is today.
	/// - Returns: Text such as "Resets at 2:00 PM" or "Resets Sunday 12:00 AM",
	///   and an empty string when there is no reset time.
	static func reset(_ resetsAt: Date?, now: Date = Date()) -> String {
		guard let resetsAt else {
			return ""
		}
		let moment = rounded(resetsAt)
		if Calendar.current.isDate(moment, inSameDayAs: now) {
			return "Resets at \(clock(resetsAt))"
		}
		let day = DateFormatter()
		day.locale = Locale(identifier: "en_US_POSIX")
		day.dateFormat = "EEEE"
		return "Resets \(day.string(from: moment)) \(clock(resetsAt))"
	}

	/// Return the lines set under a window's name.
	///
	/// - Parameters:
	///   - metric: The metric to describe.
	///   - now: Reference moment, for deciding whether the reset is today.
	/// - Returns: The reset sentence, preceded for a per model window by what
	///   that limit covers.
	static func subtitle(_ metric: Metric, now: Date = Date()) -> String {
		let resets = reset(metric.resetsAt, now: now)
		guard !metric.scopeName.isEmpty else {
			return resets
		}
		let scope = "Separate weekly limit for \(metric.scopeName)"
		return resets.isEmpty ? scope : "\(scope)\n\(resets)"
	}

	/// Return the line saying how long until the next reading.
	///
	/// - Parameter seconds: Time remaining until the next reading.
	/// - Returns: Text such as "Next update in 42s".
	static func nextUpdate(in seconds: TimeInterval) -> String {
		"Next update in \(Int(max(0, seconds)))s"
	}

	/// Return a moment rounded to the nearest minute.
	///
	/// Reset times arrive one second short of the hour, so rounding is what
	/// turns "Saturday 11:59:59 PM" into the "Sunday 12:00 AM" a reader expects.
	///
	/// - Parameter moment: The moment to round.
	/// - Returns: The same moment with its seconds dropped or carried.
	private static func rounded(_ moment: Date) -> Date {
		Date(timeIntervalSinceReferenceDate: (moment.timeIntervalSinceReferenceDate / 60).rounded() * 60)
	}
}
